### PR TITLE
(MODULES-10379) Multi Gem testing

### PR DESCRIPTION
**Pull request by sheenaajay**
Request for a review.
This PR provides feature to perform the testing of multi-gem (puppet-module-gem)
1.Cloning the gem
2.Building and pushing changes to gemfury account for testing
3. Update Gemfile with the new gem built.
Testcases are added to pdksync_spec & util_spec files.
Updated the readme with instructions on the new rake task.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -122,11 +122,42 @@ bundle install --path .bundle/gems/
 bundle exec rake git:clone_managed_modules
 bundle exec rake 'pdksync:gem_file_update[]'
 bundle exec rake 'pdksync:run_tests_locally[]'
-bundle exec rake 'pdksync:fetch_test_results[]'
+bundle exec rake 'pdksync:fetch_test_results_locally[]'
 ```
 
 Once the verified gem is released we can use pdksync to update the the new version of gem released in the  .sync.yaml file.
 
+pdksync tool is extended with the new feature to perform multi gem testing.(`puppet-module-gems`)This functionality will identify the current version and bump the version by one.Then it will build and push the gems to gemfury account.Export the GEMFURY_TOKEN to use this rake task.
```
[the remark this answers is not in the document]
Comment (sheenaajay) [reply README.md]: @cmccrisken-puppet .Sure.

File: lib/pdksync.rb [hunk 2]
```diff
@@ -87,192 +107,244 @@ def self.main(steps: [:clone], args: nil)
...
-        if module_args[:option].nil?
-          pid = Utils.run_command(output_path, module_args[:command], module_args[:option])
-          next unless pid != 0
-        else
-          exit_status = Utils.run_command(output_path, module_args[:command], nil)
-          next unless exit_status.zero?
-        end
-      end
-      if steps.include?(:gem_file_update)
+        exit_status = Utils.run_command(output_path, "sed s/current_gem_version/new_gem_version/g #{gem_args[:version_file]} >> test.yml", nil)
```
[the remark this answers is not in the document]
Comment (sheenaajay) [reply lib/pdksync.rb]: Yeah pdksync doesnt support on windows. Readme is already updated.
https://github.com/puppetlabs/pdksync/blob/master/README.md
Compatibility
This tool has been developed and tested on OSX and Linux. It currently does not run on Windows.

File: lib/pdksync.rb [hunk 3]
```diff
@@ -87,192 +107,244 @@ def self.main(steps: [:clone], args: nil)
...
-        PdkSync::Logger.info 'Fetch test results for local run '
-        module_type = Utils.module_type(output_path, module_name)
-        table = Utils.fetch_test_results_locally(output_path, module_type, module_name, report_rows)
+        Dir.chdir("#{output_path}/#{gem_args[:gem_path]}") unless Dir.pwd == output_path
+        gemfury_token = Utils.configuration.gemfury_access_settings
+        Dir.glob('*.gem') do |filename|
+          next if @gemfury_token.nil?
+          PdkSync::Logger.info filename
+          Dir.chdir(main_path) unless Dir.pwd == main_path
+          exit_status = Utils.run_command("#{output_path}/#{gem_args[:gem_path]}", "curl -F package=@#{filename} https://" + gemfury_token + "@push.fury.io/#{gem_args[:gemfury_username]}/", nil)
```
[the remark this answers is not in the document]
Comment (sheenaajay) [reply lib/pdksync.rb]: https://github.com/puppetlabs/pdksync/blob/master/README.md
pdksync doesnt support on windows now.

File: lib/pdksync/utils.rb [hunk 4]
```diff
@@ -907,5 +914,73 @@ def self.analyse_jenkins_report(url, module_name, build_id)
...
+    # @param [String] output_path
+    #   The location that the command is to be run from.
+    # @param [String] gem_name
+    #   The gem name
+    # @param [String] gemfury_token
+    #   The gemfury token
+    # @param [String] gemfury_user
+    #   The gemfury user
+    def self.update_gemfile_multigem(output_path, gem_name, gemfury_token, gemfury_user)
+      gem_file_name = 'Gemfile'
```
[the remark this answers is not in the document]
Comment (sheenaajay) [reply lib/pdksync/utils.rb]: just wondering since its applicable to only this function?

File: README.md [hunk 5]
```diff
@@ -143,7 +174,7 @@ The following rake tasks are available with pdksync:
 - `pdksync[:additional_title]` Run full pdksync process, clone repository, pdk update, create pr. Additional information can be added to the title, which will be appended before the reference section.
   - `rake pdksync` PR title outputs as `pdksync - pdksync_heads/master-0-gabccfb1`
   - `rake 'pdksync[MODULES-8231]'` PR title outputs as `pdksync - MODULES-8231 - pdksync_heads/master-0-gabccfb1`
-- `pdksync:run_a_command[:command]` Run a command against modules eg rake 'pdksync:run_a_command[complex command here -f -gx]'
+- `pdksync:run_a_command[:command, :option]` Run a command against modules eg rake 'pdksync:run_a_command[complex command here -f -gx, 'backgroud']'. :option is an optional parameter which states to run command in backgroud.
```
Comment (carabasdaniel) [inline README.md]: Typo in backgroud instead of background

File: README.md [hunk 6]
```diff
@@ -159,6 +190,9 @@ The following rake tasks are available with pdksync:
   - eg rake 'pdksync:run_tests_jenkins[test_branch, '', test_user]'
 - `pdksync:test_results_jenkins[]` Fetch traditional modules jenkins run results
   - eg rake 'pdksync:test_results_jenkins[jenkins_server_url]'
+- `git:clone_gem[:gem_name]` Clone gem.
+- `pdksync:multi_gem_testing[:gem_name, :version_file, :build_gem, :gem_path, :gemfury_user_name]` Build and Push new gems built to the gemfury account for testing eg rake 'pdksync:multi_gem_testing[]'
+- `pdksync:multigem_file_update[:gem_name, :gemfury_username]` Update Gemfile of the modules with the new gem pushed to fury.'
```
Comment (carabasdaniel) [inline README.md]: should be pushed to Gemfury

File: lib/pdksync/rake_tasks.rb [hunk 7]
```diff
@@ -38,7 +38,7 @@
     PdkSync.main(steps: [:pdk_validate])
   end
 
-  desc "Run a command against modules eg rake 'run_a_command[complex command here -f -gx]'"
+  desc "Run a command against modules eg rake 'run_a_command[complex command here -f -gx, 'backgroud']'"
```
Comment (carabasdaniel) [inline lib/pdksync/rake_tasks.rb]: backgroud typo

File: lib/pdksync/utils.rb [hunk 8]
```diff
@@ -343,6 +343,13 @@ def self.create_filespace
       configuration.pdksync_dir
     end
 
+    # @summary
+    #   This method when called will create a directory identified by the set global variable 'configuration.pdksync_gem_dir', on the condition that it does not already exist.
+    def self.create_filespace_gem
+      FileUtils.mkdir_p configuration.pdksync_dir unless Dir.exist?(configuration.pdksync_gem_dir)
```
Comment (carabasdaniel) [inline lib/pdksync/utils.rb]: it is creating the pdksync_dir, not the pdksync_gem_dir

File: lib/pdksync/utils.rb [hunk 9]
```diff
@@ -907,5 +914,73 @@ def self.analyse_jenkins_report(url, module_name, build_id)
...
+    #   The gemfury user
+    def self.update_gemfile_multigem(output_path, gem_name, gemfury_token, gemfury_user)
+      gem_file_name = 'Gemfile'
+      gem_source_line = "source \"https://#{gemfury_token}@gem.fury.io/#{gemfury_user}/\""
+      Dir.chdir(output_path) unless Dir.pwd == output_path
+
+      if gem_name.nil? == false && gemfury_token.nil? == false && gemfury_user.nil? == false # rubocop:disable Style/GuardClause
+        # Append the gem with new source location
+        gem_name = gem_name.chomp('"').reverse.chomp('"').reverse
+        File.open('/tmp/out.tmp', 'w') do |out_file|
```
Comment (carabasdaniel) [inline lib/pdksync/utils.rb]: Should probably add error handling for file operations here or where this is used

File: README.md (hunk 5 above)
Comment (sheenaajay) [reply README.md]: @carabasdaniel sure thank you.

File: lib/pdksync/utils.rb (hunk 8 above)
Comment (sheenaajay) [reply lib/pdksync/utils.rb]: @carabasdaniel Hi Dany ,all gems will be cloned to separate directory  pdksync_gem_dir

Comment (sheenaajay) [reply lib/pdksync/utils.rb]: oh got it. Changed the code.

File: lib/pdksync/utils.rb (hunk 9 above)
Comment (sheenaajay) [reply lib/pdksync/utils.rb]: thank you.adding them

**Pull request merged**